src/dbnode/encoding/proto/common.go

```diff
@@ -167,57 +177,83 @@ type encoderBytesFieldDictState struct {
 	// by comparing the bytes against those we already wrote into the
 	// stream.
 	hash     uint64
-	startPos int
-	length   int
+	startPos uint32
+	length   uint32
 }
 
-func newCustomFieldState(fieldNum int, fieldType customFieldType) customFieldState {
-	s := customFieldState{fieldNum: fieldNum, fieldType: fieldType}
-	if isUnsignedInt(fieldType) {
+func newCustomFieldState(
+	fieldNum int,
+	protoFieldType dpb.FieldDescriptorProto_Type,
+	customFieldType customFieldType,
+) customFieldState {
+	s := customFieldState{
+		fieldNum:       fieldNum,
+		fieldType:      customFieldType,
+		protoFieldType: protoFieldType}
+	if isUnsignedInt(customFieldType) {
 		s.intEncAndIter.unsigned = true
 	}
 	return s
 }
 
 // TODO(rartoul): Improve this function to be less naive and actually explore nested messages
 // for fields that we can use our custom compression on: https://github.com/m3db/m3/issues/1471
-func customAndProtoFields(s []customFieldState, protoFields []int32, schema *desc.MessageDescriptor) ([]customFieldState, []int32) {
+func customAndNonCustomFields(customFields []customFieldState, nonCustomFields []marshalledField, schema *desc.MessageDescriptor) ([]customFieldState, []marshalledField) {
 	fields := schema.GetFields()
 	numCustomFields := numCustomFields(schema)
-	numProtoFields := len(fields) - numCustomFields
+	numNonCustomFields := len(fields) - numCustomFields
 
-	if cap(s) >= numCustomFields {
-		for i := range s {
-			s[i] = customFieldState{}
+	if cap(customFields) >= numCustomFields {
+		for i := range customFields {
```

Comment:
I specifically tried to avoid pointers since I can have a lot of these in memory at once so I'd rather not have the G.C chasing them.

In term of allocating, the for loop here basically boils down to a memclear since all of the (non pointer objects) will exist as one contiguous slice in memory and its just rolling through all of them and zeroing them out. Will not call into any `malloc` or G.C code

Reply:
afaict the GC is going to end up chasing most of these anyway because they're going to be promoted from stack to heap storage as the lifetime of non-garbage `customField` members outlives their creating scope by being stored in a heap-allocated container - so the question really becomes whether it's worth it to shave off the extra stack allocations (which are admittedly not very expensive, though it's more a question of frequency/volume here) for ostensibly no change in GC behavior.

am i missing something? do you think that the difference is negligible or, worse, actually detrimental?

Reply:
I guess what I'm saying is:

```golang
[]*customFieldState
```

in memory will be represented as a contiguous region in memory where each slot is 8 bytes (enough for a 64 bit pointer to the actual customFieldState which can be anywhere else on the heap.

```golang
[]customFieldState
```

will be a contiguous region in memory where each slot is equal to `unsafe.Sizeof(customFieldState{}`

If we have a slice of size 20, in both cases the G.C is going to have to do some work because the `customFieldState` itself has pointers that need to be chase, however, the latter doesn't have any pointers for the `customFieldState` which means thats 20 less heap allocated objects that the G.C needs to trace.

Basically by making it a slice of non-pointers you're forcing them all to be next to each other in memory and their are no long-lived pointers to any individual customFieldState.

As a result, when you loop through the slice and do `nonCustomFields[i] = customFieldState{}` you're not allocating anything new. You're just going to the slot in the slice and setting the space thats already been aside to a zero instance of the struct but nothing new had to be allocated from the heap, the space was already there, you just zero'd out all the fields.

Reply:
ah, yep, you're right. i was glossing over the contiguous/non-contiguous member tradeoff. thanks for explaining!

Reply:
```golang
var dontOptimize []customFieldState

func BenchmarkResetCustomFieldState(b *testing.B) {
	dontOptimize = make([]customFieldState, 100)
	for i := 0; i < b.N; i++ {
		for i := range dontOptimize {
			dontOptimize[i] = customFieldState{}
		}
	}
	// Make sure it cant be optimized.
	fmt.Println(dontOptimize)
}
```

```
goos: darwin
goarch: amd64
pkg: github.com/m3db/m3/src/dbnode/encoding/proto
Running tool: /usr/local/bin/go test -benchmem -run=^$ github.com/m3db/m3/src/dbnode/encoding/proto -bench ^(BenchmarkResetCustomFieldState)$

BenchmarkResetCustomFieldState-8   	 2000000	       621 ns/op	       0 B/op	       0 allocs/op
PASS
ok  	github.com/m3db/m3/src/dbnode/encoding/proto	1.953s
```